hack/update-knative-components.go

```diff
@@ -0,0 +1,282 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"os"
+	"os/exec"
+	"time"
+
+	github "github.com/google/go-github/v68/github"
+)
+
+var (
+	fileScript = "hack/component-versions.sh"
+	fileJson   = "hack/component-versions.json"
+)
+
+// all the components that are kept up to date
+type Versions struct {
+	Serving  string
+	Eventing string
+	Contour  string
+}
+
+const versionsScriptTemplate = `#!/usr/bin/env bash
+
+# AUTOGENERATED FILE - edit versions in ./component-versions.yaml.
+# If you are adding components, modify this scripts' template in
+# ./update-knative-components.go
+
+set_versions() {
+	# Note: Kubernetes Version node image per Kind releases (full hash is suggested):
+	# https://github.com/kubernetes-sigs/kind/releases
+	kind_node_version={{.KindNode}}
+
+	// gets updated programatically via workflow -> PR creation
+	knative_serving_version="{{.Serving}}"
+	knative_eventing_version="{{.Eventing}}"
+	contour_version="{{.Contour}}"
+}
+`
+
+// entry function -- essentially "func main() for this file"
+func updateComponentVersions(ctx context.Context) error {
+	prTitle := "chore: Update components' versions to latest"
+	client := github.NewClient(nil).WithAuthToken(os.Getenv("GITHUB_TOKEN"))
+
+	e, err := prExists(ctx, client, prTitle)
+	if err != nil {
+		return err
+	}
+	if e {
+		fmt.Printf("PR already exists, nothing to do, exiting")
+		return nil
+	}
+
+	projects := []struct {
+		owner, repo string
+	}{
+		{
+			owner: "knative",
+			repo:  "serving",
+		},
+		{
+			owner: "knative",
+			repo:  "eventing",
+		},
+		{
+			owner: "knative-extensions",
+			repo:  "net-contour",
+		},
+	}
+
+	// Get current versions used.
+	v, err := readVersions(fileJson)
+	if err != nil {
+		return err
+	}
+
+	updated := false
+	// cycle through all versions of components listed above, fetch their
+	// latest from github releases - cmp them - create PR for update if necessary
+	for _, p := range projects {
+		newV, err := getLatestVersion(ctx, client, p.owner, p.repo)
+		if err != nil {
+			return fmt.Errorf("error while getting latest v of %s/%s: %v", p.owner, p.repo, err)
+		}
+
+		// try to overwrite with possibly new versions
+		if updateVersion(&v, p.repo, newV) {
+			// if any of the files are updated, set true
+			updated = true
+		}
```

Comment:
This is quite pedantic, but what the hay!

It's generally suggested as a Go idiom to return updated structures rather than mutate a reference unless there's a good reason for a ptr.

for example we `myslice = append(myslice, something)` rather than `append(&myslice, something)`.  Like so:

```suggestion
	v, updated = updateVersion(v, p.repo, newV)
```
(and the `if` can be collapsed into an assignment)

Reply:
ok and here I thought doing it "without the extra return" would be the better way :smile: 
thanks